src/obfuscator.cpp

```diff
@@ -23,51 +27,148 @@ bool match(re2::RE2 &regex, std::string_view value)
+{
+    bool redact_highlight = false;
+    for (auto &arg : match.args) {
+        bool sensitive_key = false;
+        for (const auto &key : arg.key_path) {
+            if (is_sensitive_key(key)) {
+                sensitive_key = true;
+                break;
             }
         }
-    }
 
-    if (!value_regex_str.empty()) {
-        const re2::StringPiece sp(value_regex_str.data(), value_regex_str.size());
-        value_regex = std::make_unique<re2::RE2>(sp, options);
+        if (sensitive_key) {
+            // The key is sensitive, we must replace the resolved value and highlight
+            redact_highlight = (arg.name == "params" || arg.name == "input");
 
-        if (!value_regex->ok()) {
-            DDWAF_ERROR("invalid obfuscator value regex: {}", value_regex->error_arg());
+            arg.resolved = redaction_msg;
+        } else {
+            if (obfuscate_value(arg.resolved)) {
+                redact_highlight = (arg.name == "params" || arg.name == "input");
+            }
         }
     }
+
+    // Since the highlight may be partial, there's no guarantee that the
+    // regular expressions will correctly match, therefore we fully redact
+    // them
+    if (redact_highlight) {
+        for (auto &highlight : match.highlights) { highlight = redaction_msg; }
+    }
 }
 
 bool obfuscator::is_sensitive_key(std::string_view key) const
 {
-    return key_regex ? match(*key_regex, key) : false;
+    return (key_regex_ && key_regex_->ok()) ? match(*key_regex_, key) : false;
 }
 
 bool obfuscator::is_sensitive_value(std::string_view value) const
 {
-    return value_regex ? match(*value_regex, value) : false;
+    return (value_regex_ && value_regex_->ok()) ? match(*value_regex_, value) : false;
+}
+
+bool obfuscator::obfuscate_value(dynamic_string &value) const
+{
+    if (!partial_value_obfuscation_) {
+        if (is_sensitive_value(value)) {
+            value = redaction_msg;
+            return true;
+        }
+        return false;
+    }
+
+    dynamic_string output{value.size()};
+    re2::StringPiece input{value.data(), value.size()};
+
+    auto prev = input;
+    while (!input.empty()) {
+        std::array<re2::StringPiece, 8> matches;
+        if (!find_and_consume(&input, *value_regex_, matches)) {
+            break;
+        }
+
+        for (auto &match : matches) {
+            if (!match.empty()) {
+                output.append({prev.data(), static_cast<std::size_t>(match.data() - prev.data())});
+                output.append(redaction_msg);
+
+                const auto *next = match.data() + match.size();
+                if (next != input.data()) {
+                    output.append({next, static_cast<std::size_t>(input.data() - next)});
+                }
+                break;
+            }
+        }
+        prev = input;
+    }
+
+    if (!output.empty()) {
+        output.append({prev.data(), prev.size()});
+        value = output;
+    }
```

Comment:
I think this could be simplified. You can lose the third append inside the loop and consolidate it with the first append.

```
char *orig_read = value.data;
while (!input.empty()) {
// ...
output.append(read, match.data() - read);
output.append(redaction_msg);
read = match.data() + match.size();
}
output.append(read, value.data() + value.size() - read);
```

Reply:
Yeah, good point, thanks